### PR TITLE
Add SymbolicDistribution and Censored distributions

### PR DESCRIPTION
**Pull request by ricardoV94**
With https://github.com/aesara-devs/aeppl/pull/22 we can now compute the logprob of censored Distributions.

This PR adds those to PyMC:

```python
import pymc as pm

with pm.Model(rng_seeder=2021) as m:
    x = pm.Normal.dist()
    y = pm.Censored('y', x, -.1, .1)

print([float(y.eval()) for _ in range(5)])
# [0.1, 0.01990278315435156, 0.1, -0.1, -0.1]

y_val = m.rvs_to_values[y]
for y_test_val in [-.2, -.1, -.09, 0, 0.09, .1, .2]:
    print(y_test_val, m.logp({y_val: y_test_val}))
# -0.2 -inf
# -0.1 -0.7761545942065803
# -0.09 -0.9229885332046727
# 0 -0.9189385332046727
# 0.09 -0.9229885332046727
# 0.1 -0.7761545942065803
# 0.2 -inf
```

A large chunk of this PR concerns how to accommodate aeppl derived distributions with PyMC API seamlessly. To do this I started by duplicating `pymc.Distribution` into `pymc.DerivedDistribution` and changing things as necessary. Hopefully we will be able to use the same class in the end or at least refactor large chunks that are shared... 

This may not be the best way... Suggestions are welcome

What works:
- [x] Defining size/shape and observed
- [x] Evaluating model logp
- [x] Prior predictive sampling
- [x] Sampling
- [x] Posterior predictive sampling

TODO:
- [x] Transforms
- [x] Dims
- [ ] String representation
- [x] Integration with `shape_utils.maybe_resize`
- [x] fix log_likelihood in InferenceDataConverter
- [x] get_moment
- [x] tests
- [x] DOCSTRINGS
- [x] Open issue in pymc-examples to update examples involving censoring

### CODE REVIEW

File: pymc/distributions/censored.py
Comment (ricardoV94) [inline pymc/distributions/censored.py]: Check for DimShuffle here

File: pymc/distributions/distribution.py
Comment (brandonwillard) [inline pymc/distributions/distribution.py]: A description/definition of a "derived distribution" might be necessary here.

**Review by brandonwillard — DISMISSED** [review]

**Comment by twiecki** [comment]
@ricardoV94 This is becoming a blocker for time series, how much is left to do here?

**Comment by ricardoV94** [comment]
I am getting back to this one this week!

**Comment by ricardoV94** [comment]
In the meantime, suggestions for the base class name? `DerivedDistribution` was my first suggestion because `Aeppl` derives the logprob terms for these graphs, but there is not much "derivation" going on in PyMC. These are just symbolic graphs that somewhere have a RandomVariable as a node / input. Other ideas:
1. `CompositeDistribution`
2. `SymbolicDistribution`
3. ... ?

**Comment by twiecki** [comment]
`CustomDistribution`? Otherwise I like `SymbolicDistribution`.

**Comment by brandonwillard** [comment]
What purpose are these `*Distribution` classes serving?  Remember, the original `Distribution` class in v4 was just an all static method-based stub/placeholder for (temporary) backwards compatibility.

**Comment by ricardoV94** [comment]
They serve the same goal the "standard" Distribution class is doing right now: basically manage all non distribution kwargs: shape/dims, transformed, observed; and behind the scenes magic things like value variables and rngs

**Comment by brandonwillard** [comment]
> They serve the same goal the "standard" Distribution class is doing right now: basically manage all non distribution kwargs: shape/dims, transformed, observed; and behind the scenes magic things like value variables and rngs

The base `Distribution` class was only ever supposed to make the hand-off to the `Model` class, which would then handle all the generic tasks (e.g. value variables, RNGs, shape/dims, etc.), and the rest was just for backward compatibility (e.g. the now completely unnecessary `Disribution.dist` interface).  Notice how instances of these classes are never made, and that only static/class/type-level functions are used.

By extending `Distribution` and adding more logic to those classes, it could become considerably more difficult to unwind those temporary backward compatibility-only choices, and they'll eventually become a permanent and confusing part of v4's design.

This isn't the time or place to start addressing all that, but these are the kinds of design considerations that need to go alongside changes/additions to the relevant areas of code (i.e. `Distribution`-related code).

**Comment by ricardoV94** [comment]
> By extending Distribution and adding more logic to those classes, it could become considerably more difficult to unwind those temporary backward compatibility-only choices, and they'll eventually become a permanent and confusing part of v4's design.

Yeah, we need ideas of how to get rid of the base distribution if that's desirable, but that is not the goal of this PR.

I created a bit of a replica of that base class (and did not modify or extended it) so hopefully whatever works for standard distributions will also work for these ones and won't be made more difficult because of the changes introduced in this PR.

File: pymc/distributions/distribution.py
Comment (twiecki) [inline pymc/distributions/distribution.py]:
```suggestion
            parameters. This may not always be constant, for instance if the symbolic
```

File: pymc/distributions/distribution.py
Comment (twiecki) [inline pymc/distributions/distribution.py:434]: Doc string is super clear, a short example would really drive the point home.

Comment (twiecki) [reply pymc/distributions/distribution.py:434]: Alternatively can point to `Censored`.

Comment (ricardoV94) [reply pymc/distributions/distribution.py:434]: Added a reference to Censored

**Review by twiecki — APPROVED** [review]

**Comment by ricardoV94** [comment]
@brandonwillard do you have other pointers for this PR? It is currently blocked by your old requested changes concerning docstrings

**Review by brandonwillard — COMMENTED** [review]
> @brandonwillard do you have other pointers for this PR? It is currently blocked by your old requested changes concerning docstrings

Sorry, I dismissed the old review; otherwise, I don't have any more input for this PR right now.

**Comment by ghost** [comment]
Thanks @ricardoV94! Can't wait to try these out.

File: pymc/distributions/distribution.py
Comment (twiecki) [inline pymc/distributions/distribution.py:358]: Will those show up in the docs?

Comment (twiecki) [reply pymc/distributions/distribution.py:358]: We also didn't mention it in the release notes.

Comment (ricardoV94) [reply pymc/distributions/distribution.py:358]: They show in the docs (you can check the RTD job in this PR). I didn't see the SymbolicDist as being user facing, that's why it's not mentioned in the release notes.

Comment (twiecki) [reply pymc/distributions/distribution.py:358]: I agree, but it's such an important new addition that I think it deserves some more attention.

**Pull request merged**
